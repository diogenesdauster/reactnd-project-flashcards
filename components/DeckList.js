import React, { PureComponent } from "react";
import {
  FlatList,
  TouchableOpacity,
  View,
  Text,
  ActivityIndicator,
  StyleSheet,
  Platform
} from "react-native";
import { lightblue3, white } from "../utils/colors";
import Deck from "./Deck";
import Ionicons from "@expo/vector-icons/Ionicons";
import PropTypes from "prop-types";
import { clearLocalNotification, setLocalNotification } from "../utils/helpers";

class DeckList extends PureComponent {
  static propTypes = {
    decks: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        questions: PropTypes.array.isRequired
      }).isRequired
    ),
    getAllDecks: PropTypes.func.isRequired,
    navigation: PropTypes.shape({
      navigate: PropTypes.func.isRequired
    })
  };
  state = { ready: false };

  componentDidMount() {
    const { getAllDecks } = this.props;
    clearLocalNotification().then(setLocalNotification);
    getAllDecks(() => this.setState({ ready: true }));
  }

  _onPress = () => {
    const { navigation } = this.props;
    navigation.navigate("CreateDeck");
  };

  _keyExtractor = item => item.title;

  _renderItem = ({ item }) => {
    const { navigation } = this.props;
    return (
      <Deck
        title={item.title}
        cardsCount={item.questions.length}
        navigation={navigation}
      />
    );
  };

  _listHeaderComponent = () => (
    <TouchableOpacity onPress={this._onPress} style={styles.container}>
      <View style={styles.button}>
        <Text style={styles.buttonText}>Create a New Deck</Text>
      </View>
    </TouchableOpacity>
  );

  _listEmptyComponent = () => {
    return (
      <View style={styles.center}>
        <Ionicons
          name={Platform.OS === "ios" ? "ios-happy-outline" : "md-happy"}
          size={100}
        />
        <Text>No Decks available, please create a New Deck</Text>
      </View>
    );
  };

  render() {
    const { decks } = this.props;
    const { ready } = this.state;

    if (!ready) {
      return <ActivityIndicator style={{ marginTop: 30 }} />;
    }

    return (
      <FlatList
        data={decks}
        ListHeaderComponent={this._listHeaderComponent}
        ListEmptyComponent={this._listEmptyComponent}
        keyExtractor={this._keyExtractor}
        renderItem={this._renderItem}
      />
    );
  }
}

const styles = StyleSheet.create({
  container: {
    marginLeft: 30,
    marginRight: 30,
    marginTop: 17,
    justifyContent: "center"
  },
  button: {
    alignItems: "center",
    backgroundColor: lightblue3
  },
  buttonText: {
    padding: 20,
    color: white
  },
  center: {
    marginTop: 50,
    justifyContent: "center",
    alignItems: "center",
    marginRight: 30,
    marginLeft: 30
  }
});

export default DeckList;
