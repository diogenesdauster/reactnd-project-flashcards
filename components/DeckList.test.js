import "react-native";
import React from "react";
import DeckList from "./DeckList";
// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";

// static propTypes = {
//     decks: PropTypes.arrayOf(
//       PropTypes.shape({
//         title: PropTypes.string.isRequired,
//         questions: PropTypes.array.isRequired
//       }).isRequired
//     ),
//     getAllDecks: PropTypes.func.isRequired,
//     navigation: PropTypes.shape({
//       navigate: PropTypes.func.isRequired
//     })
//   };

it("renders DeckList correctly", () => {
  const getAllDecks = jest.fn();
  const mockNavigation = {
    navigate: jest.fn()
  };
  const mockDecks = [
    {
      title: "React",
      questions: [
        {
          question: "What is React?",
          answer: "A library for managing user interfaces"
        },
        {
          question: "Where do you make Ajax requests in React?",
          answer: "The componentDidMount lifecycle event"
        }
      ]
    },
    {
      title: "JavaScript",
      questions: [
        {
          question: "What is a closure?",
          answer:
            "The combination of a function and the lexical environment within which that function was declared."
        }
      ]
    }
  ];

  const tree = renderer
    .create(
      <DeckList
        navigation={mockNavigation}
        decks={mockDecks}
        getAllDecks={getAllDecks}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
