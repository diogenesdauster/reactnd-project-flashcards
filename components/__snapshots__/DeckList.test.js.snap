// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DeckList correctly 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="small"
  style={
    Object {
      "marginTop": 30,
    }
  }
/>
`;
